src/GraphQL.Tests/Subscription/SubscriptionExecutionStrategyTests.cs

```diff
@@ -0,0 +1,603 @@
+            => new ObservableSelect<string, MyWidget>(source, value => new MyWidget { Name = value });
+
+        public static IObservable<MyWidget> TestNullHandling([FromServices] IObservable<string> source)
+            => new ObservableSelect<string, MyWidget>(source, value => value == null ? null! : new MyWidget { Name = value });
+
+        public static IObservable<MyWidget?> TestNullHandlingNullable([FromServices] IObservable<string> source)
+            => new ObservableSelect<string, MyWidget?>(source, value => value == null ? null : new MyWidget { Name = value });
+
+        public static IObservable<string> TestObservableNull() => null!;
+
+        public static string NotSubscriptionField() => "testing";
+    }
+
+    private class MyWidget
+    {
+        [Id]
+        public string Id => "SampleId";
+
+        public string Name { get; set; } = null!;
+
+        public string NameWithExtension(IResolveFieldContext context)
+        {
+            context.SetOutputExtension("alpha", Name);
+            return Name;
+        }
+
+        public string NameMayThrowError()
+        {
+            if (Name == "custom")
+                throw new InvalidOperationException("Custom error");
+            if (Name == "application")
+                throw new ApplicationException("Custom error");
+            return Name;
+        }
+
+        public string? NameMayThrowErrorNullable()
+        {
+            if (Name == "custom")
+                throw new InvalidOperationException("Custom error");
+            if (Name == "application")
+                throw new ApplicationException("Custom error");
+            return Name;
+        }
+
+        public int GetCounter([FromUserContext] IDictionary<string, object?> userContext)
+        {
+            var testClass = (SubscriptionExecutionStrategyTests)userContext["testClass"]!;
+            return testClass.Counter;
+        }
+
+        public bool ValidateServiceProvider(IResolveFieldContext context)
+        {
+            var provider = (IServiceProvider)context.UserContext["provider"]!;
+            return object.ReferenceEquals(context.RequestServices, provider);
+        }
+    }
+    #endregion
+
+    #region - Test helper methods and classes -
+    public class SampleListener : DocumentExecutionListenerBase
+    {
+        private readonly Action<IExecutionContext> _beforeFunc;
+        private readonly Action<IExecutionContext> _afterFunc;
+
+        public SampleListener(Action<IExecutionContext> beforeFunc, Action<IExecutionContext> afterFunc)
+        {
+            _beforeFunc = beforeFunc;
+            _afterFunc = afterFunc;
+        }
+
+        public override async Task BeforeExecutionAsync(IExecutionContext context) => _beforeFunc(context);
+
+        public override async Task AfterExecutionAsync(IExecutionContext context) => _afterFunc(context);
+    }
+
+    public class SampleObserver : IObserver<ExecutionResult>
```

Comment:
Also used in `ShouldlyExtensions` so worth to move in its own file.

Reply:
Wouldn't be surprised if there was code in System.Reactive to replace this class too.  Again, I just don't know all the classes and functionality therein.

Reply:
No. This is a custom class intended to be used by tests. It defines a specific behavior needed for tests, no infrastructure code here.

Reply:
Rx = very-very-very "smart" observables (producers)+ very-very-very "stupid" observers (consumers). Rx has thoushands lines of code jumping around data flow from one observable to another. Observer is an end of that "production chain", at least if we are talking about such observer as one above, i.e. not some internal tricky transient observer like `Subject` that plays like observer and observable at the same time. 

For me, all this is in a novelty, although I understand the common goals of all this infrastructure. Honestly, most of the Rx implementation code frustrates me with its complexity.

Reply:
Moved to its own file

Reply:
Done